// simpletts/lib/main.js

"use strict";

// deps

	// natives
	const { exec, spawn } = require("child_process");
	const { join } = require("path");

	// locals
	const parseVoicesEspeak = require(join(__dirname, "parseVoicesEspeak.js"));
	const parseVoicesSAPI = require(join(__dirname, "parseVoicesSAPI.js"));
	const readOptionsToEspeakArgs = require(join(__dirname, "readOptionsToEspeakArgs.js"));
	const readOptionsToSAPIArgs = require(join(__dirname, "readOptionsToSAPIArgs.js"));

// consts

	const IS_WINDOWS = "win32" === require("os").platform().trim().toLowerCase();
	const CSCRIPT_ARGS = [ "//NoLogo", "//B" ];

	const LISTVOICES = join(__dirname, "..", "batchs", "listvoices.vbs");
	const TTSWAV = join(__dirname, "..", "batchs", "ttswav.vbs");

// module

module.exports = class {

	constructor () {

		this._forceStop = false;

		this._writePromise = null;
		this._writing = null

		this.defaultVoice = null;
		this.forceEspeak = false;

	}

	getTTSSystem () {
		return IS_WINDOWS && !this.forceEspeak ? "sapi" : "espeak";
	}

	getVoices () {

		const IS_SAPI = "sapi" === this.getTTSSystem();

		return new Promise((resolve, reject) => {

			exec(IS_SAPI ?
				"cscript " + CSCRIPT_ARGS.join(" ") + " \"" + LISTVOICES + "\"" :
				"espeak --voices",
			(err, _stdout, stderr) => {

				if (err) {
					reject(stderr);
				}
				else {

					const stdout = _stdout.trim().replace(/\r/g, "\n").replace(/\n\n/g, "\n").split("\n");

					resolve(
						IS_SAPI ?
							parseVoicesSAPI(stdout.slice(1, stdout.length), stdout) :
							parseVoicesEspeak(stdout.slice(1, stdout.length))
					);

				}

			});

		// standardise
		}).then((voices) => {

			const result = [];

				voices.forEach((_voice) => {
					const voice = {
						"gender": IS_SAPI ? _voice.Gender ? _voice.Gender.trim().toLowerCase() : "" : "",
						"name": IS_SAPI ? _voice.Name.trim().toLowerCase() : _voice.VoiceName.trim().toLowerCase()
					};

					if (!IS_SAPI) {
						voice.gender = "F" === _voice["Age/Gender"] ? "female" : "male";
					}

					result.push(voice);

				});

			return Promise.resolve(result);

		});

	}



	ttswav (_options) {

		this._forceStop = false;

		if ("undefined" === typeof _options) {
			return Promise.reject(new ReferenceError("Missing options parameter"));
		}
			else if ("object" !== typeof _options && "string" !== typeof _options) {
				return Promise.reject(new TypeError("options parameter is not an object or a string"));
			}

		else {
			const options = "string" === typeof _options ? { "text": _options } : _options;

			if ("undefined" === typeof options.text) {
				return Promise.reject(new ReferenceError("Missing text parameter"));
			}
			else if ("string" !== typeof options.text) {
				return Promise.reject(new TypeError("text parameter is not a string"));
			}
			else if ("" === options.text.trim()) {
				return Promise.reject(new Error("text parameter is empty"));
			}
			else if ("undefined" === typeof options.file) {
				return Promise.reject(new ReferenceError("Missing file parameter"));
			}
			else if ("string" !== typeof options.file) {
				return Promise.reject(new TypeError("file parameter is not a string"));
			}
			else if ("" === options.file.trim()) {
				return Promise.reject(new Error("file parameter is empty"));
			}
			else if (this.isWriting()) {
				return Promise.reject(new Error("Already writing a text"));
			}
			else {

				const IS_SAPI = "sapi" === this.getTTSSystem();

				this._writePromise = Promise.resolve().then(() => {

					return this.defaultVoice ? Promise.resolve(this.defaultVoice) : this.getVoices().then((voices) => {
						this.defaultVoice = voices.shift(); return Promise.resolve();
					});

				}).then(() => {

					if ("object" !== typeof options.voice && "string" !== typeof options.voice) {
						options.voice = this.defaultVoice;
					}

					options.volume = "undefined" === typeof options.volume ? 100 : Math.round(options.volume);
						options.volume = 0 > options.volume ? 0 : options.volume;
						options.volume = 100 < options.volume ? 100 : options.volume;

					options.speed = "undefined" === typeof options.speed ? 50 : Math.round(options.speed);
						options.speed = 0 > options.speed ? 0 : options.speed;
						options.speed = 100 < options.speed ? 100 : options.speed;

					return this._forceStop ? Promise.resolve(options) : Promise.resolve().then(() => {

						const args = readOptionsToSAPIArgs(options);

						args.push(options.text);

						return Promise.resolve(args);

					}).then((args) => {

						return new Promise((resolve, reject) => {
							this._writing = spawn("cscript", CSCRIPT_ARGS.concat([ TTSWAV ]).concat(args));

							let err = "";
							this._writing.stderr.on("data", (data) => {
								err += "string" === typeof data ? data : data.toString("ascii");
							});

							this._writing.on("close", (code) => {
								return code ? reject(new Error(err)) : resolve(options);
							});

						});

					});

				}).then((data) => {

					this._forceStop = false;
					this._writing = null;

					return Promise.resolve(data);

				}).catch((err) => {

					this._forceStop = false;
					this._writing = null;

					return Promise.reject(err);

				});

				return this._writePromise;

			}

		}

	}

	isWriting () {
		return null !== this._writing;
	}

};
